linera-service/src/cli_wrappers/wallet.rs

```diff
@@ -1698,3 +1714,112 @@ impl<A> From<String> for ApplicationWrapper<A> {
+
+    /// Waits for a `NewIncomingBundle` notification for the given sender chain and sender block
+    /// height. If no height is specified, any height is accepted.
+    fn wait_for_bundle(
+        &mut self,
+        origin: ChainId,
+        height: impl Into<Option<BlockHeight>>,
+    ) -> impl Future<Output = Result<()>>;
+}
+
+#[cfg(with_testing)]
+impl<T: Stream<Item = Result<Notification>>> NotificationsExt for Pin<Box<T>> {
+    async fn wait_for_events(
+        &mut self,
+        expected_height: impl Into<Option<BlockHeight>>,
+    ) -> Result<BTreeSet<StreamId>> {
+        let expected_height = expected_height.into();
+        let mut timeout = Box::pin(linera_base::time::timer::sleep(notification_timeout())).fuse();
+        loop {
+            let notification = futures::select! {
+                () = timeout => bail!("Timeout waiting for notification"),
+                notification = self.next().fuse() => notification.context("Stream closed")??,
+            };
+            if let Reason::NewEvents {
+                height,
+                event_streams,
+                ..
+            } = notification.reason
+            {
+                if expected_height.is_none_or(|h| h == height) {
+                    return Ok(event_streams);
+                }
+            }
+        }
+    }
+
+    async fn wait_for_block(
+        &mut self,
+        expected_height: impl Into<Option<BlockHeight>>,
+    ) -> Result<CryptoHash> {
+        let expected_height = expected_height.into();
+        let mut timeout = Box::pin(linera_base::time::timer::sleep(notification_timeout())).fuse();
+        loop {
+            let notification = futures::select! {
+                () = timeout => bail!("Timeout waiting for notification"),
+                notification = self.next().fuse() => notification.context("Stream closed")??,
+            };
+            if let Reason::NewBlock { height, hash, .. } = notification.reason {
+                if expected_height.is_none_or(|h| h == height) {
+                    return Ok(hash);
+                }
+            }
+        }
+    }
+
+    async fn wait_for_bundle(
+        &mut self,
+        expected_origin: ChainId,
+        expected_height: impl Into<Option<BlockHeight>>,
+    ) -> Result<()> {
+        let expected_height = expected_height.into();
+        let mut timeout = Box::pin(linera_base::time::timer::sleep(notification_timeout())).fuse();
+        loop {
+            let notification = futures::select! {
+                () = timeout => bail!("Timeout waiting for notification"),
+                notification = self.next().fuse() => notification.context("Stream closed")??,
+            };
+            if let Reason::NewIncomingBundle { height, origin } = notification.reason {
+                if expected_height.is_none_or(|h| h == height) && origin == expected_origin {
+                    return Ok(());
+                }
+            }
```

Comment:
All three functions have almost the same implementation except for:
- `Reason` variant they await for
- the subsequent check

nit: maybe there's a clean way to extract it to a function and just pass the closure that takes `notification` and returns `Ok(T)`.